bigc/exceptions.py

```diff
@@ -1,83 +1,181 @@
 from __future__ import annotations
 
-from typing import Optional
+import inspect
+from typing import Any, ClassVar
 
-from requests import Response
+import requests
 
+_status_code_exc_class_map: dict[int, type[BigCommerceException]] = {}
 
-class BigCommerceAPIException(Exception):
-    """Base exception class for BigCommerce API errors"""
 
-    def __init__(self, message: Optional[str] = None, response: Optional[Response] = None):
-        self._message = message
-        self.response = response
+class BigCommerceException(Exception):
+    """Base exception class for BigCommerce API errors."""
 
-        super().__init__(
-            message or f'Request to BigCommerce failed{f" with {self.message}" if response is not None else ""}'
-        )
+    DEFAULT_MESSAGE: ClassVar[str] = 'Request to BigCommerce failed.'
+    STATUS_CODE: ClassVar[int | None] = None
 
-    @property
-    def status_code(self) -> Optional[int]:
-        if self.response is None:
-            return None
+    def __init__(self,
+                 message: str | None = None,
+                 response: Any = None,
+                 errors: dict[str, str] | None = None):
+        """
+        :param message: A short description of the error's cause.
+        :param response: The underlying API response, if any, that caused the
+            error. Currently, this is a ``requests.Response``, but this is an
+            implementation detail that's subject to change.
+        :param errors: A dictionary that may contain additional error details.
+        """
+        super().__init__(message, response, errors)
 
-        return self.response.status_code
+        self.message = message
+        self.response = response
+        self.errors = errors
 
     @property
-    def message(self) -> str:
-        if self._message is not None:
-            return self._message
+    def status_code(self) -> int | None:
+        """The HTTP status code of the response, if one was provided."""
+        if isinstance(self.response, requests.Response):
+            return self.response.status_code
+        return None
```

Comment:
I almost wonder if this should be passed in separately (similar to `message` and `errors`)?

Reply:
So that our exceptions don't have to be aware of the underlying library we're using for HTTP requests? We're already upholding that with us typing `response` as `Any` and not calling `.json()` in this file right?

Reply:
1ff9f44